### ISSUE TITLE
Z3 returns unexpected results for specific formula

### ISSUE THREAD

**Issue by giladreich** [Z3Prover/z3#4919]
Giving the following formula:
`((p → q) ∧ (r → s) ∧ (p ∨ ¬ s)) ├ (q ∨ ¬ r)`

The following code should describe it:
```
(declare-const p Bool)
(declare-const q Bool)
(declare-const r Bool)
(declare-const s Bool)
(assert (=> (and (and (=> p q) (=> r s)) (or p (not s))) (or q (not r))))
(check-sat)
(get-model)
```

Expected results for `get-model` should be:
```
(model 
  (define-fun r () Bool
    true)
)
```

But returned:
```
(
  (define-fun p () Bool
    false)
  (define-fun q () Bool
    false)
  (define-fun s () Bool
    false)
  (define-fun r () Bool
    false)
)
```

It returns the unexpected results from the latest version of Z3 using the `master` branch:
```
❯ z3 -version
Z3 version 4.8.10 - 64 bit
```

However, for version:
```
❯ z3 -version
Z3 version 4.6.0 - 64 bit
```

Everything works as expected.

**Comment by andrewvaughanj** [comment]
Why do you think that model is "unexpected"?

On the left-hand side of the entailment you have a conjunction of three clauses, all of which can be satisfied by assigning the propositional variables to `false` -- and `false` on the LHS of an implication is always `true` (overall).

If you're trying to _prove_ the formula (vs. check for satisfiability and finding a single model), then you probably need some quantifiers.

**Comment by giladreich** [comment]
Hi @andrewvaughanj. Thank you very much for the input.

> Why do you think that model is "unexpected"?

I was expecting to receive output of only one declared variable and also `r` to be `true` based on the results I received from Z3 v4.6.0 - regardless of the version differences. I understand that there could be multiple solutions to satisfy this formula, but I thought it was a bug as I installed Z3 from source using the `master` branch and thought it would be nice reporting.

I figured out that the behavior has changed from the older Z3 I was using. e.g. in the older version for some reason it will output all variables using this specific formula **only** if I break it down into multiple assertions. But for some reason when using one big assertion and calling `get-model`, it only showed the state of the variable used by the last clause, which is `r`.

I understand now that it's not a bug, rather something that improved with time.
